src/common/src/array/stream_chunk_builder.rs

```diff
@@ -132,39 +132,221 @@ impl StreamChunkBuilder {
     #[must_use]
     pub fn append_record(&mut self, record: Record<impl Row>) -> Option<StreamChunk> {
         match record {
-            Record::Insert { new_row } => self.append_row(Op::Insert, new_row),
-            Record::Delete { old_row } => self.append_row(Op::Delete, old_row),
+            Record::Insert { new_row } => {
+                self.append_iter_inner::<true>(Op::Insert, new_row.iter().enumerate())
+            }
+            Record::Delete { old_row } => {
+                self.append_iter_inner::<true>(Op::Delete, old_row.iter().enumerate())
+            }
             Record::Update { old_row, new_row } => {
-                let none = self.append_row(Op::UpdateDelete, old_row);
-                debug_assert!(none.is_none());
-                self.append_row(Op::UpdateInsert, new_row)
+                let none =
+                    self.append_iter_inner::<true>(Op::UpdateDelete, old_row.iter().enumerate());
+                assert!(none.is_none());
+                self.append_iter_inner::<true>(Op::UpdateInsert, new_row.iter().enumerate())
             }
         }
     }
 
+    /// Take all the pending data and return a chunk. If there is no pending data, return `None`.
+    /// Note that if this is an unlimited chunk builder, the only way to get a chunk is to call
+    /// `take`.
     #[must_use]
     pub fn take(&mut self) -> Option<StreamChunk> {
         if self.size == 0 {
             return None;
         }
-
         self.size = 0;
-        let new_columns = self
+
+        let ops = std::mem::replace(&mut self.ops, Vec::with_capacity(self.initial_capacity));
+        let columns = self
             .column_builders
             .iter_mut()
             .zip_eq_fast(&self.data_types)
             .map(|(builder, datatype)| {
-                std::mem::replace(builder, datatype.create_array_builder(self.capacity)).finish()
+                std::mem::replace(
+                    builder,
+                    datatype.create_array_builder(self.initial_capacity),
+                )
+                .finish()
             })
             .map(Into::into)
             .collect::<Vec<_>>();
-
         let vis = std::mem::take(&mut self.vis_builder).finish();
 
-        Some(StreamChunk::with_visibility(
-            std::mem::replace(&mut self.ops, Vec::with_capacity(self.capacity)),
-            new_columns,
-            vis,
-        ))
+        Some(StreamChunk::with_visibility(ops, columns, vis))
+    }
+
+    #[must_use]
+    fn append_iter_inner<'a, const VIS: bool>(
+        &mut self,
+        op: Op,
+        iter: impl IntoIterator<Item = (usize, DatumRef<'a>)>,
```

Comment:
There's no dynamic dispatch here, I think the performance should be the same as `for (builder, datum) in builders.zip_eq(row.iter()) { .. }`

Reply:
I don't know. Both `iter()` and passing by value here have some extra overhead. Since this call is definitely a critical code path, every CPU cycle matters.

Reply:
> I think the performance should be the same as `for (builder, datum) in builders.zip_eq(row.iter()) { .. }`.

+1. I guess there won't be too much overhead. 🤔

Reply:
Trust you ❤️